components/brave_wallet_ui/common/async/lib.ts

```diff
@@ -163,75 +165,177 @@ export async function getIsSwapSupported (network: BraveWallet.NetworkInfo): Pro
+        ...token,
+        logo: token.symbol.toLowerCase() === 'sol' ? '' : `chrome://erc-token-images/${token.logo}`,
+        chainId: network.chainId
+      })) as BraveWallet.BlockchainToken[]
+
+      return tokenList.length === 0 ? [nativeAsset] : tokenList
+    }))
+    const visibleAssets = getVisibleAssets.flat(1)
     await dispatch(WalletActions.setVisibleTokensInfo(visibleAssets))
   }
 }
 
 export function refreshBalances () {
   return async (dispatch: Dispatch, getState: () => State) => {
     const { jsonRpcService } = getAPIProxy()
-    const { wallet: { accounts, userVisibleTokensInfo, selectedNetwork } } = getState()
+    const { wallet: { accounts, userVisibleTokensInfo, networkList, defaultNetworks } } = getState()
+
+    const emptyBalance = {
+      balance: '0x0',
+      error: 0,
+      errorMessage: ''
+    }
+
+    const getNativeAssetsBalanceReturnInfos = await Promise.all(accounts.map(async (account) => {
+      const networks = GetNetworksByCoinType(networkList, account.coin)
+
+      // Get CoinType SOL balances
+      if (account.coin === BraveWallet.CoinType.SOL) {
+        return Promise.all(networks.map(async (network) => {
+          // MULTICHAIN: Backend needs to allow chainId for getSolanaBalance method
+          if (defaultNetworks.some(n => n.chainId === network.chainId)) {
+            const solBalanceInfo = await jsonRpcService.getSolanaBalance(account.address)
+            return {
+              ...solBalanceInfo,
+              balance: solBalanceInfo.balance.toString(),
+              chainId: network.chainId
+            }
+          }
+          return {
+            ...emptyBalance,
+            chainId: network.chainId
+          }
+        }))
+      }
 
-    const getBalanceReturnInfos = await Promise.all(accounts.map(async (account) => {
-      const balanceInfo = await jsonRpcService.getBalance(account.address, account.coin, selectedNetwork.chainId)
-      return balanceInfo
+      // Get CoinType FIL balances
+      if (account.coin === BraveWallet.CoinType.FIL) {
+        return Promise.all(networks.map(async (network) => {
+          // MULTICHAIN: Backend needs to allow different chainId's
+          // for FIL for the getBalance method
+          if (defaultNetworks.some(n => n.chainId === network.chainId)) {
+            const balanceInfo = await jsonRpcService.getBalance(account.address, account.coin, network.chainId)
+            return {
+              ...balanceInfo,
+              chainId: network.chainId
+            }
+          }
+          return {
+            ...emptyBalance,
+            chainId: network.chainId
+          }
+        }))
+      }
+
+      // Get CoinType ETH balances
+      return Promise.all(networks.map(async (network) => {
+        // LOCALHOST will return an error until a local instance is
+        // detected, we now will will return a 0 balance until it's detected.
+        if (network.chainId === BraveWallet.LOCALHOST_CHAIN_ID) {
```

Comment:
suggesting that we move the Solana and Filecoin logic in here so we only have to maintain one promise mapping

Reply:
Combined into 1 promise mapping for all `coinTypes`